src/main/java/fr/insalyon/citi/golo/compiler/ParseTreeToGoloIrVisitor.java

```diff
@@ -198,65 +199,62 @@ public Object visit(ASTFunctionDeclaration node, Object data) {
         node.getName(),
         node.isLocal() ? LOCAL : PUBLIC,
         node.isAugmentation() ? AUGMENT : MODULE);
+    if (node.isDecorator()) {
+      function.setDecorator(true);
+    }
     node.setIrElement(function);
     while (context.objectStack.peek() instanceof Decorator) {
-      function.addDecorator((Decorator)context.objectStack.pop());
+      function.addDecorator((Decorator) context.objectStack.pop());
     }
     context.objectStack.push(function);
     if (!function.getDecorators().isEmpty()) {
-      ASTFunction original = (ASTFunction)node.jjtGetChild(0);
-      Node wrapped = wrapFunctionWithDecorators(original, function.getDecorators());
-      wrapped.jjtAccept(this, data);
-    } else {
-      node.childrenAccept(this, data);
+      function.setDecoratorRef("__$$_" + function.getName() + "_decorator_" + context.nextDecoratorId++);
```

Comment:
maybe you should use the decorator name (if available) in the synthetic function name (more readable when debugging). 
Else, I think `__$$_decorated_foo_0` is a better name.